src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs

```diff
@@ -418,5 +390,105 @@ public void DebugVerifyEmpty()
             }
 #endif
         }
+
+        // A note about the IOThreadScheduler and the ScheduledOverlapped references:
+        // Although for each scheduler we have a single instance of overlapped, we cannot point to the scheduler from the
+        // overlapped, through the entire lifetime of the overlapped. This is because the ScheduledOverlapped is pinned
+        // and if it has a reference to the IOTS, it would be rooted and the finalizer will never get called.
+        // Therefore, we are passing the reference, when we post a pending callback and reset it, once the callback was
+        // invoked; during that time the scheduler is rooted but in that time we don't want that it would be collected
+        // by the GC anyway.
+        private unsafe class ScheduledOverlapped
+        {
+            private static bool s_isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+            private readonly NativeOverlapped* _nativeOverlapped;
+            private IOThreadScheduler _scheduler;
+            private Action _postDelegate;
+
+            public ScheduledOverlapped()
+            {
+                if (s_isWindows)
+                {
+                    _nativeOverlapped = (new Overlapped()).UnsafePack(
+                        Fx.ThunkCallback(new IOCompletionCallback(IOCallback)), null);
+                    _postDelegate = PostIOCP;
+                }
+                else
+                {
+                    _postDelegate = PostNewThread;
```

Comment:
I think we should experiment with this being a QieueUserWorkItem, but that can be a discussion for another day :-)

Reply:
I feel like it can’t really hurt but we can leave it